### PR TITLE
Update Deprecated :simple_one_to_one to use DynamicSupervisor in Elixir 1.6

### PR DESCRIPTION
**Pull request by rockerBOO**
* Use dynamic supervision with DynamicSupervisor. 
https://github.com/elixir-lang/elixir/blob/v1.6/CHANGELOG.md#dynamic-supervisor
https://hexdocs.pm/elixir/DynamicSupervisor.html#init/1
https://hexdocs.pm/elixir/DynamicSupervisor.html#module-migrating-from-supervisor-s-simple_one_for_one

* Add ExIRC.start_link!/1

**Requires Elixir 1.6**

### CODE REVIEW

**Comment by tchoutri** [comment]
@rockerBOO Hi, and thank you for this PR! Does it change anything regarding the public API? Could you include an example that leverages the dynamic supervision?

**Comment by rockerBOO** [comment]
`test "client dies if owner process dies"` tests this behavior in [test/client_test.exs](https://github.com/bitwalker/exirc/blob/master/test/client_test.exs#L14). 

`ExIRC.start_client!` will attach a client to the `:exirc` supervisor. This can help to make sure the clients die when the main `:exirc` supervisor dies. We add `restart: :temporary` because we are OK with the client dying in this case.  This should replicate old behavior. 

`Supervisor.Spec` was also deprecated.

>Supervisor.Spec
>This module is deprecated. Use the new child specifications outlined in the Supervisor module instead.

>The functions in this module are deprecated and they do not work with the module-based child specs introduced in Elixir v1.5. Please see the Supervisor documentation instead.

We were using the `:simple_one_for_one` supervisor, which was also deprecated. The [suggestion](https://hexdocs.pm/elixir/DynamicSupervisor.html#module-migrating-from-supervisor-s-simple_one_for_one) is to move to a  `DynamicSupervisor`. 

## Added

To the public API, to allow access to the new module-based child specs, I added `ExIRC.start_link!/1` which allows us to do `{ExIRC, []}` to start the application. We only had `ExIRC.start_link!/0`. I am dropping the values in the currently but that replicates current behavior. Maybe we should pass options down?

**Comment by rockerBOO** [comment]
The actual problem case Im having, which might require expanding this (or another pull). 

In my tests I want to make a named client that is [start_supervised](https://hexdocs.pm/ex_unit/ExUnit.Callbacks.html#start_supervised/2).

```elixir
  test "connection to the Twitch server" do
    client = start_supervised!(ExIRC)

    Process.register(client, :twitchirc)
  end
```
This fails because no `ExIRC.start_link/1` since `start_supervised!(ExIRC)` will default to `start_supervised!({ExIRC, []})`

### What I want

What I want to do is like the following. Pass a name opts argument to the client. 

```elixir
  test "connection to the Twitch server" do
    client = start_supervised!({ExIRC, [], name: :twitchirc})
  end
```

That would mean something like

```elixir
defmodule ExIRC do 
    @spec start_link!(opts \\ [], process_opts \\ []) 
```

**Comment by tchoutri** [comment]
@rockerBOO You can make that addition to this PR. Also, make sure to rebase your branch on the current `master`.

File: lib/exirc/exirc.ex [hunk 1]
```diff
@@ -29,8 +29,24 @@ defmodule ExIRC do
       ExIRC.Client.stop! client
 
   """
-  use Supervisor
-  import Supervisor.Spec
+  use DynamicSupervisor
+
+  defmodule TemporaryClient do
```
Comment (rockerBOO) [inline lib/exirc/exirc.ex:34]: This allows us to define the module differently if it's started using the dynamic supervisor. There may be a few ways to do this but this seemed the most clear currently.

**Comment by rockerBOO** [comment]
So after doing some testing I learned that I just needed `{ExIRC.Client, []}` or `ExIRC.Client.start_link/1`. We also already have `ExIRC.Client.start_link/2` so I can already pass process `arg` to this. So I removed `ExIRC.start_link!/1`.

So the key points of the pull request is now **only with dynamic supervision to replace the deprecated code.**

Let me know if there is something I messed up. The `.yaml` file seems weird to be adding for example.

File: README.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
 # ExIRC
 
 [![Build Status](https://travis-ci.org/bitwalker/exirc.svg?branch=master)](https://travis-ci.org/bitwalker/exirc)
+![.github/workflows/tests.yaml](https://github.com/rockerBOO/exirc/workflows/.github/workflows/tests.yaml/badge.svg)
```
Comment (tchoutri) [inline README.md:4]: Beware, you're using your own repo's status for this badge

Comment (rockerBOO) [reply README.md]: Good spot :) thanks.

**Comment by tchoutri** [comment]
@rockerBOO I don't see anything to add. I'll let you aggregate your commits if this PR is ready to be merged. It's a fine work altogether, thank you very much for your patience and efforts.

**Comment by rockerBOO** [comment]
OK Travis, Github green. Should be good now. 

We likely want to increment the number as the version requirement has changed and will cause breaking due to `DynamicSupervisor` added in 1.6.

Thanks Théophile! Thank you for your help :)

**Comment by tchoutri** [comment]
Ah, it would seem that Coveralls is unhappy. Do you think you can make it smile again?

**Comment by rockerBOO** [comment]
Ok coveralls is happy.

**Comment by tchoutri** [comment]
Great, merging it now.

**Pull request merged**